### PR TITLE
fix: docs generator

### LINKED ISSUE #14820: bug(codegen): docs generator fails when nested module has same name as function or resource
### What happened?

As described in https://github.com/pulumi/pulumi/pull/14308#issuecomment-1774812922 and https://github.com/pulumi/pulumi/pull/14308#issuecomment-1777972676 the docs generator in `codegen/docs/gen.go` has bug that leads to a panic when a module has the same name as a resource or function within it. 

The `moduleConflictResolver` should prevnt this to happen but it does not.

```go
panic: fatal: An assertion has failed: duplicate file: router/bgp/_index.md

goroutine 1 [running]:
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
        /home/vagrant/code/pulumi/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assertf(0x30?, {0x1ae9498?, 0xc00d629530?}, {0xc000bcf040?, 0x48797e?, 0x1a24940?})
        /home/vagrant/code/pulumi/sdk/go/common/util/contract/assert.go:35 +0xed
github.com/pulumi/pulumi/pkg/v3/codegen.Fs.Add(...)
        /home/vagrant/code/pulumi/pkg/codegen/utilities.go:180
github.com/pulumi/pulumi/pkg/v3/codegen/docs.(*modContext).gen.func1({0x0, 0x0}, {0x1adb5af, 0xa}, {0x1a24940, 0xc00e3303c0})
        /home/vagrant/code/pulumi/pkg/codegen/docs/gen.go:1854 +0x225
github.com/pulumi/pulumi/pkg/v3/codegen/docs.(*modContext).gen(0xc000bf2aa0, 0xc000bede00)
        /home/vagrant/code/pulumi/pkg/codegen/docs/gen.go:1962 +0x11db
github.com/pulumi/pulumi/pkg/v3/codegen/docs.(*docGenContext).generatePackage(0xc0000c1400, {0x0?, 0x0?}, 0x0?)
        /home/vagrant/code/pulumi/pkg/codegen/docs/gen.go:2201 +0x2e8
github.com/pulumi/pulumi/pkg/v3/codegen/docs.GeneratePackage(...)
        /home/vagrant/code/pulumi/pkg/codegen/docs/gen.go:2255
github.com/pulumi/registrygen/pkg.generateDocsFromSchema({0x7ffeb02ae50a, 0xf}, 0xc4a256?)
        /home/vagrant/code/pulumi-registrygen/pkg/docs.go:218 +0x54
github.com/pulumi/registrygen/pkg.GenerateDocs({0x7ffeb02ae493?, 0x0?}, {0x7ffeb02ae4b8, 0x6}, {0x7ffeb02ae4cc, 0x30}, {0x7ffeb02ae50a, 0xf}, {0x7ffeb02ae532, 0xb}, ...)
        /home/vagrant/code/pulumi-registrygen/pkg/docs.go:111 +0x896
github.com/pulumi/registrygen/cmd/docs.PackageDocsCmd.func1(0xc0007ff800?, {0x1ad0743?, 0x9?, 0x9?})
        /home/vagrant/code/pulumi-registrygen/cmd/docs/cli.go:89 +0x74
github.com/spf13/cobra.(*Command).execute(0xc0007ff800, {0xc0006943f0, 0x9, 0x9})
        /home/vagrant/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc0007fe600)
        /home/vagrant/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(0xc000124120?)
        /home/vagrant/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992 +0x19
main.main()
        /home/vagrant/code/pulumi-registrygen/main.go:30 +0x8b
```

### Example

**Example:**

Module: `Router` Resource: `Bgp`: path for `resource.tmpl` file = `router/bgp/_index.md`
Module: `Router/Bgp`: path for module _index.md file = `router/bgp/_index.md`

### Output of `pulumi about`

The `registrygen` tool is build using 

```
github.com/pulumi/pulumi/pkg/v3 v3.95.0
github.com/pulumi/pulumi/sdk/v3 v3.95.0
```

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### LINKED ISSUE #14821: bug(codegen): docs generator creates invalid package tree for nested modules
### What happened?

The `generatePackageTree` in `codegen/docs/package_tree.go` creates an invalid package tree when nested modules are used inside the provider schema to separate resources and functions into namespaces of the supported programming languages.

Invalid tree

```json
  {
    "name": "log",
    "type": "module",
    "link": "log/",
    "children": [
      {
        "name": "v2",
        "type": "module",
        "link": "log/fortianalyzer/v2/",
        "children": [
          {
            "name": "Filter",
            "type": "resource",
            "link": "res-filter"
          },
          {
            "name": "Overridefilter",
            "type": "resource",
            "link": "res-overridefilter"
          },
          {
            "name": "Overridesetting",
            "type": "resource",
            "link": "res-overridesetting"
          },
          {
            "name": "Setting",
            "type": "resource",
            "link": "res-setting"
          }
        ]
      },

```

Expected tree

```json
  {
    "name": "log",
    "type": "module",
    "link": "log/",
    "children": [
      {
        "name": "syslogd",
        "type": "module",
        "link": "log/syslogd/",
        "children": [
          {
            "name": "v2",
            "type": "module",
            "link": "log/syslogd/v2/",
            "children": [
              {
                "name": "Filter",
                "type": "resource",
                "link": "res-filter"
              },
              {
                "name": "Overridefilter",
                "type": "resource",
                "link": "res-overridefilter"
              },
              {
                "name": "Overridesetting",
                "type": "resource",
                "link": "res-overridesetting"
              },
              {
                "name": "Setting",
                "type": "resource",
                "link": "res-setting"
              }
            ]
          }
```

### Example

Snippet from the provider schema:

```json
        "fortios:log/syslogd/v2/filter:Filter": {

        },
        "fortios:log/syslogd/v2/overridefilter:Overridefilter": {

        },
        "fortios:log/syslogd/v2/overridesetting:Overridesetting": {

        },
        "fortios:log/syslogd/v2/setting:Setting": {

        },
```

### Output of `pulumi about`

The `registrygen` tool is build using 

```
github.com/pulumi/pulumi/pkg/v3 v3.95.0
github.com/pulumi/pulumi/sdk/v3 v3.95.0
```

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by tmeckel**
# Description

This PR contains fixes to solve two issues with the docs generator.

Fixes #14820
Fixes #14821

To fix #14820 the moduleConflictResolver has been completely removed from the docs generator and has been replaced with a strategy to always add a prefix `res-` to a directory for a resource and a `fn-` prefix for a directory of a function. This will effectively prevent name collisions without the overhead of moduleConflictResolver which tried to solve this problem in some dynamic way.

The new implementation has been tested by using `registrygen gen docs` with the following Pulumi providers: Azure Classic, GCP, AWS, Docker, Kubernetes and the result yielded only changes to the folder names of all resources and functions, what was expected.

To fix #14821 the code at `getMod` has been changed so that `parentName := path.Dir(modName)` has been removed and the `modName` is passed directly to the recursive call of `dctx.getMod` because `pkg.TokenToModule` always removes the last element of the module name and thus `path.Dir(modName)` prematurely shortened the path so that the recursive call of `dctx.getMod`  didn't return the parent module but the next but one module in the hierarchy.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

### CODE REVIEW

**Comment by t0yv0** [comment]
This seems reasonable, what I worry about here is this, are the folder names for documentation free-floating or is there something significant depending on them, such as back-links or URLs in Pulumi registry.

**Comment by t0yv0** [comment]
I think we have a way to preview registry changes for a PR, let me dig that up and see if this can be stood up for say pulumi-aws for review.

**Comment by tmeckel** [comment]
> This seems reasonable, what I worry about here is this, are the folder names for documentation free-floating or is there something significant depending on them, such as back-links or URLs in Pulumi registry.

@t0yv0  I think it would be useful if @cnunciato would also give his opinion on this PR, as I understood that he's the maintainer of the docs platform.

**Comment by tmeckel** [comment]
> I think we have a way to preview registry changes for a PR, let me dig that up and see if this can be stood up for say pulumi-aws for review.

@t0yv0 as I stated in the initial comment I only verified that `registrgen` is able to create the docs of the mostly used and biggest Pulumi providers. Because I really struggled with the Pulumi Hugo setup in the past, I didn't tested the complete publishing process.

**Comment by t0yv0** [comment]
Hmm

"""
/run-docs-gen - used to test the Pull Request against documentation generation
"""

I've not tried this but perhaps worth a try. Let's try this first.

**Comment by tmeckel** [comment]
@t0yv0 awesome. Issue that I've to change/update ALL tests in the docs package before this PR can merged. I haven't done this yet because I first wanted to get an opinion on whether the PR can be merged at all.

**Comment by tmeckel** [comment]
> /run-docs-gen
@t0yv0 Perhaps the PR must be ready for review. Still in draft mode.

**Comment by t0yv0** [comment]
Reading the job for this I think it's bit-rotted, I'm not seeing it create registry PRs anyway. 

So creating manually for now: 

https://github.com/pulumi/registry/pull/3708

File: pkg/codegen/docs/gen.go [hunk 1]
```diff
@@ -2026,12 +1957,10 @@ func (dctx *docGenContext) getMod(
 		}
 
 		if modName != "" && codegen.PkgEquals(tokenPkg, pkg) {
-			parentName := path.Dir(modName)
 			// If the parent name is blank, it means this is the package-level.
-			if parentName == "." || parentName == "" {
+			parentName := ":" + modName + ":"
```
Comment (t0yv0) [inline pkg/codegen/docs/gen.go:1961]: This is causing an infinite loop on pulumi-google-native

```
(*docGenContext) getMod(
  token = "google-native:accesscontextmanager/v1:AccessLevel"
  tokenPkg = packageDefRef{pkg: Name: "google-native"}
  pkg = packageDefRef{pkg: Name: "google-native" }
  tool = "Pulumi Docs Generator"
  add = true
)
```

subsequently it loops with parentName = ":accesscontextmanager/v1:"

**Comment by t0yv0** [comment]
http://registry--origin-pr-3708-61e44067.s3-website.us-west-2.amazonaws.com/registry/packages/akamai/api-docs/res-appsecadvancedsettingsprefetch/

vs

https://www.pulumi.com/registry/packages/akamai/api-docs/appsecadvancedsettingsprefetch/

Unfortunately this "res-" prefix leaks into registry URLs which means ALL registry pages will move with this PR breaking any external links to these pages. As far as I could tell Hugo does not automatically create redirects for these. 

Copying @cnunciato to confirm but I think this might be a big problem. If that is indeed a blocker we might need to find how to fix the bugs without changing the existing URLs drastically.

**Comment by cnunciato** [comment]
> Copying @cnunciato to confirm but I think this might be a big problem. If that is indeed a blocker we might need to find how to fix the bugs without changing the existing URLs drastically.

Thanks for the head's up, and you're right, this wouldn't be handled by Hugo automatically. We'd need to some sort of redirect plan in place and ready to go before we could merge this. There are a few ways we could do that, and it'd take a bit of work, but it's definitely doable.

**Comment by tmeckel** [comment]
> > Copying @cnunciato to confirm but I think this might be a big problem. If that is indeed a blocker we might need to find how to fix the bugs without changing the existing URLs drastically.
> 
> Thanks for the head's up. Yeah, you're right -- this wouldn't be handled by Hugo automatically. We'd need some sort of redirect plan in place and ready to go before we could merge this. There are a few ways we could do that, and it'd take a bit of work, but it's definitely doable, and something we'd probably need to do anyway to fix the current issue, at least as I understand it.

@cnunciato I worked together with @t0yv0 on this and he came up with a much cleaner approach which will not require any changes to the urls. In addition issue #14820 has been solved implicitly by fixing the code which generates the package tree #14821. Thus I'll close this PR here in favor of https://github.com/pulumi/pulumi/pull/15035

**Comment by t0yv0** [comment]
Still working out testing for https://github.com/pulumi/pulumi/pull/15035 but will finish soon-ish. Thank you!

**Pull request closed without merging**